File: apps/saas-admin-portal/src/services/analyticsService.ts
import { api } from './api';

// 定义流程统计类型
export interface ProcessStats {
  tenantId?: string;
  tenantName?: string;
  templateId?: string;
  templateName?: string;
  totalExecutions: number;
  successfulExecutions: number;
  failedExecutions: number;
  avgDuration: number; // 平均执行时间(秒)
  successRate: number; // 成功率(%)
  periodStart: string;
  periodEnd: string;
}

// 定义模板使用情况类型
export interface TemplateUsage {
  templateId: string;
  templateName: string;
  category: string;
  usageCount: number;
  lastUsed: string;
  avgDuration: number; // 平均执行时间(秒)
}

// 定义租户统计类型
export interface TenantStats {
  tenantId: string;
  tenantName: string;
  totalProcesses: number;
  totalExecutions: number;
  activeProcesses: number;
  failedProcesses: number;
  periodStart: string;
  periodEnd: string;
}

// 定义错误分析类型
export interface ErrorAnalysis {
  tenantId?: string;
  tenantName?: string;
  templateId?: string;
  templateName?: string;
  errorType: string;
  errorMessage: string;
  count: number;
  firstOccurrence: string;
  lastOccurrence: string;
  affectedProcesses: string[];
}

// 定义优化建议类型
export interface OptimizationSuggestion {
  tenantId?: string;
  templateId?: string;
  suggestionType: 'performance' | 'reliability' | 'usability';
  description: string;
  priority: 'low' | 'medium' | 'high';
  estimatedImpact: string;
  implementationGuide: string;
}

// 定义分析报告类型
export interface AnalyticsReport {
  reportId: string;
  generatedAt: string;
  periodStart: string;
  periodEnd: string;
  tenantId?: string;
  summary: {
    totalProcesses: number;
    totalExecutions: number;
    successRate: number;
    avgDuration: number;
  };
  topTemplates: Array<{
    templateId: string;
    templateName: string;
    usageCount: number;
    successRate: number;
  }>;
  errorSummary: {
    totalErrors: number;
    topErrorTypes: Array<{
      errorType: string;
      count: number;
      percentage: number;
    }>;
  };
  recommendations: OptimizationSuggestion[];
}

// 定义获取分析数据的参数类型
export interface AnalyticsParams {
  tenantId?: string;
  templateId?: string;
  startDate?: string;
  endDate?: string;
}

// 分析服务
export const analyticsService = {
  // 获取流程统计信息
  getProcessStats: async (params?: AnalyticsParams): Promise<ProcessStats[]> => {
    try {
      const queryParams = new URLSearchParams();
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            queryParams.append(key, String(value));
          }
        });
      }
      
      const response = await api.get<ProcessStats[]>(`/analytics/process-stats?${queryParams.toString()}`);
      return response;
    } catch (error) {
      throw new Error(`获取流程统计信息失败: ${error}`);
    }
  },

  // 获取模板使用情况
  getTemplateUsage: async (params?: Omit<AnalyticsParams, 'templateId'>): Promise<TemplateUsage[]> => {
    try {
      const queryParams = new URLSearchParams();
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            queryParams.append(key, String(value));
          }
        });
      }
      
      const response = await api.get<TemplateUsage[]>(`/analytics/template-usage?${queryParams.toString()}`);
      return response;
    } catch (error) {
      throw new Error(`获取模板使用情况失败: ${error}`);
    }
  },

  // 获取租户统计信息
  getTenantStats: async (params?: Omit<AnalyticsParams, 'tenantId'>): Promise<TenantStats[]> => {
    try {
      const queryParams = new URLSearchParams();
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            queryParams.append(key, String(value));
          }
        });
      }
      
      const response = await api.get<TenantStats[]>(`/analytics/tenant-stats?${queryParams.toString()}`);
      return response;
    } catch (error) {
      throw new Error(`获取租户统计信息失败: ${error}`);
    }
  },

  // 获取错误分析
  getErrorAnalysis: async (params?: AnalyticsParams): Promise<ErrorAnalysis[]> => {
    try {
      const queryParams = new URLSearchParams();
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            queryParams.append(key, String(value));
          }
        });
      }
      
      const response = await api.get<ErrorAnalysis[]>(`/analytics/error-analysis?${queryParams.toString()}`);
      return response;
    } catch (error) {
      throw new Error(`获取错误分析失败: ${error}`);
    }
  },

  // 获取优化建议
  getOptimizationSuggestions: async (params?: AnalyticsParams): Promise<OptimizationSuggestion[]> => {
    try {
      const queryParams = new URLSearchParams();
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            queryParams.append(key, String(value));
          }
        });
      }
      
      const response = await api.get<OptimizationSuggestion[]>(`/analytics/optimization-suggestions?${queryParams.toString()}`);
      return response;
    } catch (error) {
      throw new Error(`获取优化建议失败: ${error}`);
    }
  },

  // 生成分析报告
  generateReport: async (params?: AnalyticsParams): Promise<AnalyticsReport> => {
    try {
      const response = await api.post<AnalyticsReport>('/analytics/generate-report', params);
      return response;
    } catch (error) {
      throw new Error(`生成分析报告失败: ${error}`);
    }
  },

  // 导出分析报告
  exportReport: async (params?: AnalyticsParams): Promise<Blob> => {
    try {
      const response = await api.post<Blob>('/analytics/export-report', params, {
        responseType: 'blob',
      });
      return response;
    } catch (error) {
      throw new Error(`导出分析报告失败: ${error}`);
    }
  },
};

File: apps/saas-api/app/services/analytics_service.py
from typing import List, Optional, Dict, Any
import uuid
import logging
from datetime import datetime, timedelta
from app.schemas.analytics import (
    ProcessStats, 
    TemplateUsage, 
    TenantStats, 
    ErrorAnalysis,
    OptimizationSuggestion
)

# 配置日志
logger = logging.getLogger(__name__)

class AnalyticsService:
    """
    分析服务类，负责处理分析相关的业务逻辑
    """
    
    def __init__(self):
        """
        初始化分析服务
        """
        # 在实际实现中，这里需要连接数据库和日志服务
        # 为了简化，我们使用内存存储
        self.process_executions = []
        self.tenants = {}
        self.templates = {}
        self.errors = []
    
    def get_process_stats(
        self,
        tenant_id: Optional[str] = None,
        template_id: Optional[str] = None,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None
    ) -> List[ProcessStats]:
        """
        获取流程统计信息
        
        Args:
            tenant_id: 租户ID筛选
            template_id: 模板ID筛选
            start_time: 开始时间筛选
            end_time: 结束时间筛选
            
        Returns:
            流程统计信息列表
        """
        # 在实际实现中，这里会从数据库查询数据
        # 为了简化，我们返回模拟数据
        
        # 默认时间范围为最近30天
        if not start_time:
            start_time = datetime.now() - timedelta(days=30)
        if not end_time:
            end_time = datetime.now()
            
        stats = ProcessStats(
            tenant_id=tenant_id,
            template_id=template_id,
            template_name="销售订单处理流程" if template_id else None,
            total_executions=100,
            successful_executions=85,
            failed_executions=15,
            avg_duration=120.5,
            success_rate=0.85,
            period_start=start_time,
            period_end=end_time
        )
        
        return [stats]
    
    def get_template_usage(
        self,
        tenant_id: Optional[str] = None,
        limit: int = 100
    ) -> List[TemplateUsage]:
        """
        获取模板使用情况
        
        Args:
            tenant_id: 租户ID筛选
            limit: 返回结果数量限制
            
        Returns:
            模板使用情况列表
        """
        # 在实际实现中，这里会从数据库查询数据
        # 为了简化，我们返回模拟数据
        
        usage = TemplateUsage(
            template_id="template_1",
            template_name="销售订单处理流程",
            tenant_id=tenant_id,
            usage_count=50,
            last_used=datetime.now(),
            avg_duration=120.5
        )
        
        return [usage]
    
    def get_tenant_stats(
        self,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None
    ) -> List[TenantStats]:
        """
        获取租户统计信息
        
        Args:
            start_time: 开始时间筛选
            end_time: 结束时间筛选
            
        Returns:
            租户统计信息列表
        """
        # 在实际实现中，这里会从数据库查询数据
        # 为了简化，我们返回模拟数据
        
        # 默认时间范围为最近30天
        if not start_time:
            start_time = datetime.now() - timedelta(days=30)
        if not end_time:
            end_time = datetime.now()
            
        stats = TenantStats(
            tenant_id="tenant_1",
            tenant_name="示例公司",
            total_processes=5,
            total_executions=200,
            active_processes=3,
            failed_processes=2,
            period_start=start_time,
            period_end=end_time
        )
        
        return [stats]
    
    def get_error_analysis(
        self,
        tenant_id: Optional[str] = None,
        template_id: Optional[str] = None,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None
    ) -> List[ErrorAnalysis]:
        """
        获取错误分析
        
        Args:
            tenant_id: 租户ID筛选
            template_id: 模板ID筛选
            start_time: 开始时间筛选
            end_time: 结束时间筛选
            
        Returns:
            错误分析列表
        """
        # 在实际实现中，这里会从日志服务查询错误数据
        # 为了简化，我们返回模拟数据
        
        analysis = ErrorAnalysis(
            tenant_id=tenant_id,
            template_id=template_id,
            error_type="参数验证错误",
            error_message="缺少必需参数: customer_id",
            count=5,
            first_occurrence=datetime.now() - timedelta(days=5),
            last_occurrence=datetime.now() - timedelta(hours=2),
            affected_processes=["process_1", "process_2"]
        )
        
        return [analysis]
    
    def get_optimization_suggestions(
        self,
        tenant_id: Optional[str] = None,
        template_id: Optional[str] = None
    ) -> List[OptimizationSuggestion]:
        """
        获取优化建议
        
        Args:
            tenant_id: 租户ID筛选
            template_id: 模板ID筛选
            
        Returns:
            优化建议列表
        """
        # 在实际实现中，这里会基于AI分析日志和执行数据生成建议
        # 为了简化，我们返回模拟数据
        
        suggestion = OptimizationSuggestion(
            tenant_id=tenant_id,
            template_id=template_id,
            suggestion_type="performance",
            description="建议为'销售订单处理流程'增加缓存机制，以减少重复查询",
            priority="medium",
            estimated_impact="可减少30%的执行时间",
            implementation_guide="在步骤2和步骤4之间添加缓存节点"
        )
        
        return [suggestion]
    
    def generate_report(
        self,
        tenant_id: Optional[str] = None,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None
    ) -> Dict[str, Any]:
        """
        生成分析报告
        
        Args:
            tenant_id: 租户ID筛选
            start_time: 开始时间筛选
            end_time: 结束时间筛选
            
        Returns:
            分析报告
        """
        # 在实际实现中，这里会生成详细的分析报告
        # 为了简化，我们返回模拟数据
        
        # 默认时间范围为最近30天
        if not start_time:
            start_time = datetime.now() - timedelta(days=30)
        if not end_time:
            end_time = datetime.now()
            
        report = {
            "report_id": str(uuid.uuid4()),
            "generated_at": datetime.now(),
            "period_start": start_time,
            "period_end": end_time,
            "tenant_id": tenant_id,
            "summary": {
                "total_processes": 5,
                "total_executions": 200,
                "success_rate": 0.85,
                "avg_duration": 120.5
            },
            "top_templates": [
                {
                    "template_id": "template_1",
                    "template_name": "销售订单处理流程",
                    "usage_count": 50,
                    "success_rate": 0.90
                }
            ],
            "error_summary": {
                "total_errors": 15,
                "top_error_types": [
                    {
                        "error_type": "参数验证错误",
                        "count": 5,
                        "percentage": 33.3
                    }
                ]
            },
            "recommendations": [
                {
                    "type": "performance",
                    "description": "建议为'销售订单处理流程'增加缓存机制",
                    "priority": "medium"
                }
            ]
        }
        
        return report
